<!doctype html>
<!-- vim: set sw=4:ts=4:expandtab:cindent: -->
<%/*
    Groovy Template for Generating CNV Report

    This template is designed to be executed by Bpipe to generate the CNV report.
    It expects the following parameters to by supplied in the model:

        -   cnv_summary : file containing tab separated columns with details
                          about the CNVs. This file is generated by R.

        -   batch_name  :  a name for the batch to which the samples belong.

        -   callers     : the names of the CNV callers. Each of these should
                          a column in the provided cnv_summary (along with many 
                          other columns)

        -   types       : List of CNV types to include (DUP,DEL)

        -   target_region : name of BED file defining the targeted regions for the 
                            capture
        -   dgv_file : DGV database, downloaded from UCSC (dgvMerged.txt)
*/%>
<%
    // Read the CNVs from the file
    // println("Processing file ${new File(cnv_summary).absolutePath}")
    def lines = new File(cnv_summary).readLines()
    // println("Processing ${lines.size()} lines")
    def headers = lines[0].split('\\s')

    def cnvTypes = types
    if(types instanceof String) {
        cnvTypes = types.split(',') as List
    }

    //println(headers)
    allCnvs = lines[1..-1].collect { 
        [headers,it.split('\\s')].transpose().collectEntries()
    }

    def chrCounts = [:]
    allCnvs.eachWithIndex { cnv, index ->
        chrCounts[cnv.seqnames] = chrCounts.containsKey(cnv.seqnames) ? chrCounts[cnv.seqnames]+1 : 1 
        cnv.chrIndex = chrCounts[cnv.seqnames]
    }

    cnvs = allCnvs.grep { it.type in cnvTypes }

    // println("Processing ${cnvs.size()} cnvs")
    //println(cnvs);
    def cnv_labels = [
        DEL : "Deletion",
        DUP : "Duplication"
    ]

    new File("template.log").withWriter { log ->

    // Find the Variants for each CNV
    def sample_vcfs = cnvs.unique(false) {it.sample}.collectEntries { cnv ->
        log.println "Raw VCF for $cnv.sample = $cnv.vcf"
	if(cnv.vcf == null)
		return [ cnv.sample, null ]
        def vcf = cnv.vcf.endsWith(".bgz") ? cnv.vcf.replaceAll('.bgz$','') : cnv.vcf
        log.println "Sample = $cnv.sample, used VCF = $vcf"
        [ cnv.sample, new VCFIndex(vcf) ]
    }

    log.println "Loading variants ..."
    def variants = cnvs.collect { cnv ->
        def vcf = sample_vcfs[cnv.sample]
        def variants = []
	if(vcf == null)
		return []
        vcf.query(cnv.seqnames, cnv.start.toInteger(), cnv.end.toInteger()) { if(it.info.DP.toInteger()>5) variants.add(it) }
        log.println "Loaded ${variants.size()} variants for cnv $cnv.seqnames:$cnv.start-$cnv.end ..."
        if(variants.size()>50)
            return []
        else
            return variants;
    }

    def anno_types = [ "DEL" : "LOSS", "DUP" : "GAIN" ]

    // Load the CNV annotation information
    BED targetRegions = new BED(target_bed).load()
    TargetedCNVAnnotator annotator = new TargetedCNVAnnotator(targetRegions, dgv_file)

    log.println "Writing report ..."
%>
<html>
    <head>
        <title>CNV Report for Batch $batch_name</title>
        <style type='text/css'>
            body {
                font-family: helvetica;
            }

            #filterHelp {
                font-size: 80%;
                font-style: italic;
                color: #888;
            }
 
            table th { text-align: left; }
            table tr td:nth-child(1) { text-align: center; }
            hr { clear: both; margin: 2em 0; }
            img.cnvimg { display : block; }
            table.cnvTable {
                font-size: 10px;
                font-family: verdana;
            }
            table.cnvTable th {
                font-weight: bold;
            }
            .ui-layout-east {
                font-size: 10px;
                font-family: verdana;
            }
        </style>
        <script type='text/javascript' src='http://ajax.aspnetcdn.com/ajax/jQuery/jquery-2.1.0.min.js'></script>
        <script type='text/javascript' src="http://cdn.datatables.net/1.10.0/js/jquery.dataTables.min.js"></script>
 
        <script src="http://ajax.googleapis.com/ajax/libs/jqueryui/1.10.4/jquery-ui.min.js"></script>
        <script type='text/javascript' src='http://cdnjs.cloudflare.com/ajax/libs/jquery-layout/1.3.0-rc-30.79/jquery.layout.min.js'></script>
        <script type='text/javascript' src='cnv.js'></script>

        <link rel="stylesheet" href='http://cdn.datatables.net/1.10.0/css/jquery.dataTables.css'>
        <link rel="stylesheet" href="//ajax.googleapis.com/ajax/libs/jqueryui/1.10.4/themes/smoothness/jquery-ui.css" />
        <link rel="stylesheet" href='http://cdnjs.cloudflare.com/ajax/libs/jquery-layout/1.3.0-rc-30.79/layout-default.css'>

        <script type='text/javascript'>
            <% log.println "Writing cnv json" %>
            var cnvs = <%=groovy.json.JsonOutput.toJson(cnvs)%>;
            <% log.println "Writing variant json" %>
            var cnvVariants = <%="[" + variants.collect{ "["+ it*.toJson().join(",") + "]" }.join(",") + "]"%>;
            <% log.println "Writing callers json" %>
            var callers = <%=groovy.json.JsonOutput.toJson(callers)%>;
            <% log.println "Writing annotations json" %>
            <% def annotations = cnvs.collect { cnv -> annotator.annotate(new Region(cnv.seqnames, (cnv.start.toInteger())..(cnv.end.toInteger())), anno_types[cnv.type])}; %>
            var annotations = <%=groovy.json.JsonOutput.toJson(annotations)%>;
        </script>

        <%if(inline_js) {%>
        <script type='text/javascript'>
            <%=new File(new File(templatePath).parentFile, 'cnv.js').text%>
        </script>
        <%} else {%>
            <script type='text/javascript' src='cnv.js'></script>
        <%}%>
    </head>
    <body>
        <div class="ui-layout-north">
            <h1>CNV Report for Batch ${batch_name}</h1>
            <input type=checkbox id='filterHVR' name='filterHVR'> Filter out high variability regions
            <span id=filterOuter><span id=filters> </span></span>
            <div id=filterHelp style='display:none'> </div>
        </div>

        <div class="ui-layout-center">
            <h2>Summary</h2>
             <div id=tableHolder>
                <table id='cnvTable' class='stripe'>
                    <thead>
                        <%
                         def tableHeaders = ['Index','Genes','Region','Size','Type','True?','Callers','Sample','Conc','Het Count', 'Hom Count', 'Balance', 'DGV Match', 'DGV Span', 'Freq']
                         if(!callers.contains("truth")) {
                            tableHeaders.remove("True?")
                         }
                        %>
                        <tr><%=tableHeaders.collect{"<th>$it</th>"}.join('')%></tr>
                    </thead>
                    <tbody>
                    <% cnvs.eachWithIndex { cnv, index -> %>
                        <% log.println "Writing cnv $index ($cnv.seqnames:$cnv.start-$cnv.end)" %>
                        <% def genes = cnv.genes.split(",") as List; 
                           if(genes.size()>5) {
                             genes = [genes[0],genes[1],"... ${genes.size()-4} genes ...",genes[-2],genes[-1]]
                           }
                        %>
                        <tr class='cnvRow' id='cnv_${index}_row'>
                            <td><a href='#cnv_${index+1}_detail'>${index+1}</a></td>
                            <td>${genes.join(",")}</td>
                            <td>${cnv.seqnames}:${cnv.start}-${cnv.end}</td>
                            <td>${cnv.end.toInteger() - cnv.start.toInteger()}</td>
                            <td>${cnv.type}</td>
                            <%if(callers.contains("truth")) {%>
                                <td><%=cnv["truth"]=="TRUE"?"Yes":"No"%></td>
                            <%}%>
                            <td><%=callers.grep { it != "truth" && cnv[it] == "TRUE" }.join(",")%></td>
                            <td>${cnv.sample}</td>
                            <td>${cnv.count}</td>
                            <td>${cnv.het}</td>
                            <td>${cnv.hom}</td>
		            <%if(cnv.bal!=null) {%>
                            <td>${cnv.bal.isFloat()?String.format("%2.3f",cnv.bal.toFloat()):cnv.bal}</td>
                            <%} else {%> <td>0</td> <%}%>
                            <td>${annotations[index].matching?.size()?:""}</td>
                            <td>${annotations[index].spanning.size()}</td>
                            <td>${String.format("%.4f",annotations[index].spanningFreq)}</td>
                        </tr>
                    <% } %>
                    </tbody>
                </table>
            </div>
            <hr>
            <h2>Coverage Diagrams</h2>
            <% cnvs.eachWithIndex { cnv, index -> %>
                <% log.println "Writing cnv img $index ($cnv.chrIndex) ($cnv.seqnames:$cnv.start-$cnv.end)" %>
                <div id='cnv_${index+1}_detail'>
                <h3 >${index+1}. ${cnv_labels[cnv.type]} at $cnv.seqnames:$cnv.start-$cnv.end over genes ${cnv.genes} in Sample ${cnv.sample}</h3>
                <div id='cnv_${index+1}_img'>
                    <img class='cnvimg' src='cnv_${cnv.seqnames}_${cnv.chrIndex}.rel.png' width=920>

                    <% /* <img class='cnvimg' src='cnv_${cnv.seqnames}_${cnv.chrIndex}.abs.png' width=920> */ %>
                </div>
                <hr>
                </div>
            <%}%>
        </div>
        <div class="ui-layout-south">Report created ${new Date()} from source ${new File(cnv_summary).absolutePath} (${lines.size()} lines)</div>
        <div class="ui-layout-east" id="ui-layout-east"></div>
    </body>
</html>
<% } /* end log writer */ %>
